ledger/error/src/main/scala/com/daml/error/definitions/IndexErrors.scala

```diff
@@ -56,21 +53,34 @@ object IndexErrors extends IndexErrorGroup {
         extends ErrorCode(
           id = "INDEX_DB_INVALID_RESULT_SET",
           ErrorCategory.SystemInternalAssumptionViolated,
-        )
-        with HasUnapply {
+        ) {
       case class Reject(message: String)(implicit
           val loggingContext: ContextualizedErrorLogger
-      ) extends LoggingTransactionErrorImpl(
+      ) extends DbError(
             cause = message
           )
     }
   }
 
-  trait HasUnapply {
-    this: ErrorCode =>
-    // TODO error codes: Create a generic unapply for ErrorCode that returns the ErrorCode instance
-    //                   and match against that one.
-    def unapply(exception: StatusRuntimeException): Option[Unit] =
-      if (ErrorDetails.isErrorCode(exception)(errorCode = this)) Some(()) else None
+  // Decorator that returns a specialized StatusRuntimeException (IndexDbException)
+  // that can be used for precise matching of persistence exceptions (e.g. for index initialization failures that need retrying).
+  // Without this specialization, internal errors just appear as StatusRuntimeExceptions (see INDEX_DB_SQL_NON_TRANSIENT_ERROR)
+  // without any marker, impeding us to assert whether they are emitted by the persistence layer or not.
+  abstract class DbError(
+      override val cause: String,
+      override val throwableO: Option[Throwable] = None,
+  )(implicit
+      code: ErrorCode,
+      loggingContext: ContextualizedErrorLogger,
+  ) extends LoggingTransactionErrorImpl(cause, throwableO) {
+    override def asGrpcErrorFromContext(implicit
+        contextualizedErrorLogger: ContextualizedErrorLogger
+    ): IndexDbException = {
+      val err = super.asGrpcErrorFromContext(contextualizedErrorLogger)
+      IndexDbException(err.getStatus, err.getTrailers)
```

Comment:
Out of interest, is `IndexDbException` name leaked to the user of the API? (I.e. can users tell a difference between getting `StatusRuntimeException` and `IndexDbException`?)

Reply:
I don't think so. What gets over the wire is the status and metadata. JVM type information should be ignored I believe

Reply:
Possibly in logs but that shouldn't be an issue I think